modules/caddytls/tls.go

```diff
@@ -164,6 +164,33 @@ func (t *TLS) Provision(ctx caddy.Context) error {
 		t.certificateLoaders = append(t.certificateLoaders, modIface.(CertificateLoader))
 	}
 
+	// on-demand permission module
+	if t.Automation != nil && t.Automation.OnDemand != nil && t.Automation.OnDemand.PermissionRaw != nil {
+		val, err := ctx.LoadModule(t.Automation.OnDemand, "PermissionRaw")
+		if err != nil {
+			return fmt.Errorf("loading on-demand TLS permission module: %v", err)
+		}
+		t.Automation.OnDemand.permission = val.(OnDemandPermission)
+	}
+
+	// on-demand rate limiting
+	if t.Automation != nil && t.Automation.OnDemand != nil && t.Automation.OnDemand.RateLimit != nil {
+		onDemandRateLimiter.SetMaxEvents(t.Automation.OnDemand.RateLimit.Burst)
+		onDemandRateLimiter.SetWindow(time.Duration(t.Automation.OnDemand.RateLimit.Interval))
+	} else {
+		// remove any existing rate limiter
+		onDemandRateLimiter.SetWindow(0)
+		onDemandRateLimiter.SetMaxEvents(0)
+	}
+
+	// run replacer on ask URL (for environment variables) -- return errors to prevent surprises (#5036)
+	if t.Automation != nil && t.Automation.OnDemand != nil && t.Automation.OnDemand.Ask != "" {
+		t.Automation.OnDemand.Ask, err = repl.ReplaceOrErr(t.Automation.OnDemand.Ask, true, true)
+		if err != nil {
+			return fmt.Errorf("preparing 'ask' endpoint: %v", err)
+		}
+	}
```

Comment:
Probably missing something here, but when reading this it looks duplicated (the `PermissionsByHTTP` module is doing it on it's endpoint already), and also a bit off: Wouldn't the parsing of the configuration already have the `ask` field consumed here, and produced a `PermissionRaw` thing?

Reply:
There's JSON config _and_ Caddyfile config to worry about. This PR changes the Caddyfile to always produce the HTTP module for backwards compatibility, but we can't control what users of JSON continue to use in their config so we need to still accept the old field in JSON and warn to let them switch over to the new way.

Reply:
Yeah, so this is a compatibility layer for the 'ask' property of the JSON config, which many people will still be using. This code as well as the 'ask' property will eventually be removed once enough people have time to migrate.

Reply:
Right, thanks! 

(And yes, I know some people that are using JSON :D)